src/Emulator/Peripherals/Peripherals/Miscellaneous/SevenSegments.cs

```diff
@@ -0,0 +1,197 @@
+                StringBuilder result = new StringBuilder();
+
+                foreach(var b in state)
+                {
+                    if(Digits.ContainsKey((byte)(b & ~SEGMENT_DOT)))
+                    {
+                        result.Append(Digits[(byte)(b & ~SEGMENT_DOT)]);
+                    }
+                    else if(b == 0)
+                    {
+                        result.Append("_");
+                    }
+                    else if(b != SEGMENT_DOT)
+                    {
+                        result.Clear();
+                        break;
+                    }
+                    if((b & SEGMENT_DOT) != 0)
+                    {
+                        result.Append(".");
+                    }
+                }
+
+                if(result.Length == 0)
+                {
+                    foreach(var b in state)
+                        result.AppendFormat("0x{0:x2} ", b);
+
+                }
+
+                return result.ToString();
+            }
+        }
+
+        public byte[] State
+        {
+            get => state;
+
+            private set
+            {
+                lock(sync)
+                {
+                    if(value == state)
+                    {
+                        return;
+                    }
+
+                    state = value;
+                    StateChanged?.Invoke(this, state);
+                    this.Log(LogLevel.Noisy, "Seven Segments state changed to {0}", Image);
+                }
+            }
+        }
+
+        private byte[] BuildImage()
+        {
+            byte segments = 0;
+            for(int index = 0; index < 8; index++)
+            {
+                if(rows[index])
+                {
+                    segments |= (byte)(1 << index);
+                }
+            }
+
+            byte[] image = new byte[digitsCount];
+            for(int index = 0; index < digitsCount; index++)
+            {
+                if(cols[index])
+                {
+                    image[index] = segments;
+                }
+                else
+                {
+                    image[index] = 0;
+                }
+            }
+            return image;
+        }
+
+        private bool[] rows;
+        private bool[] cols;
+        private byte[] state;
+
+        private readonly uint digitsCount;
+        private readonly bool invertSegments;
+        private readonly bool invertDigits;
+        private readonly object sync;
+
+        private const byte SEGMENT_A = (1 << 0);
+        private const byte SEGMENT_B = (1 << 1);
+        private const byte SEGMENT_C = (1 << 2);
+        private const byte SEGMENT_D = (1 << 3);
+        private const byte SEGMENT_E = (1 << 4);
+        private const byte SEGMENT_F = (1 << 5);
+        private const byte SEGMENT_G = (1 << 6);
+        private const byte SEGMENT_DOT = (1 << 7);
+
+        private static readonly Dictionary<byte, string> Digits = new Dictionary<byte, string>()
```

Comment:
Instead of encoding the state of a digit in a `byte` variable I would suggest adding some internal private class/struct for it with methods/properties that return a readble text and/or graphic representation of the symbol.